tools/database/src/utils.rs

```diff
@@ -49,3 +65,190 @@ pub(crate) fn resolve_column(col_name: &str) -> anyhow::Result<DBCol> {
+/// parallel_loader module for a description of the parallel loading algorithm and why these nodes
+/// are needed.
+///
+/// Also return a reasonable partitioning of the FlatState column keys.
+///
+/// Despite returning Result, for readability and ease of implementation, this function will panic
+/// for some errors that are not meaningful to recover from anyway.
+pub fn prepare_memtrie_state_trimming(
+    store: Store,
+    genesis_config: &GenesisConfig,
+    include_flat_delta: bool,
+) -> anyhow::Result<MemtrieStateTrimmingCalculationResult> {
+    let epoch_manager = EpochManager::new_arc_handle(store.clone(), genesis_config);
+    let chain = ChainStore::new(store.clone(), genesis_config.genesis_height, false);
+    let final_head = chain.final_head()?;
+    let epoch_id = final_head.epoch_id;
+    let shard_layout = epoch_manager.get_shard_layout(&epoch_id)?;
+    let shard_uids = shard_layout.shard_uids().collect::<Vec<_>>();
+
+    let mut flat_state_ranges = Vec::new();
+
+    // Spawn a thread to collect all the important keys that we'll discover.
+    let (important_keys_tx, important_keys_rx) = std::sync::mpsc::sync_channel::<Vec<u8>>(1000000);
+    let important_keys_collect_thread = std::thread::spawn(move || {
+        let mut important_keys = Vec::new();
+        while let Ok(key) = important_keys_rx.recv() {
+            important_keys.push(key);
+        }
+        important_keys
+    });
+    for shard_uid in shard_uids {
+        tracing::info!("Reading shard {shard_uid}...");
+        let flat_status = get_flat_storage_status(&store, shard_uid)?;
+        let flat_head = match flat_status {
+            near_store::flat::FlatStorageStatus::Ready(ready) => ready.flat_head.hash,
+            _ => panic!("Flat storage is not ready for shard {shard_uid}"),
+        };
+        let root = get_state_root(&store, flat_head, shard_uid)?;
+        let plan = make_memtrie_parallel_loading_plan(store.clone(), shard_uid, root)?;
+
+        // The nodes that are needed to load the memtrie must be kept in the State column.
+        for hash in plan.node_hashes() {
+            let key = encode_flat_state_db_key(shard_uid, &hash.0);
+            important_keys_tx.send(key).unwrap();
+        }
+
+        // Memtrie returns us a list of prefixes that can be used to load the memtrie in parallel.
+        // However, since these ranges exclude the leaf nodes that are in the plan, we need to tweak
+        // the ranges a bit to make sure they cover all the FlatState keys. We'll do that by simply
+        // taking into account only the start key of each prefix (ignoring the first).
+        let mut split_points = plan
+            .subtrees_to_load
+            .into_iter()
+            .map(|subtree| {
+                let (start, _) = subtree.to_iter_range(shard_uid);
+                start
+            })
+            .collect::<Vec<_>>();
+        split_points.sort();
+        split_points.remove(0); // skip first split point
```

Comment:
nit: could also just start iterating from 1 below instead of this O(n) operation? but if i understand correctly this will have around 256 elements so doesnt matter too much?

Reply:
I'm afraid of a crash though, if the whole range happens to be the empty prefix. I can also start iterating from 1 but it makes the code feel less easy to see that it's safe.

The O(n) operation is trivial anyway because at most this will have about a couple of thousand elements.